### DISCUSSION TITLE
Lesson 7 Refactoring

### DISCUSSION THREAD

**Discussion by TemoTech25** [Cyfrin/foundry-full-course-cu#125] [Q&A] [answered]
I am passing 2 tests and failing one after refactoring and can understand where i am going wrong, please do help 
![Screenshot (4)](https://github.com/Cyfrin/foundry-full-course-f23/assets/135317073/dd1c6d3d-e967-4aaa-bc93-79bcfef33fa8)

**Comment by alymurtazamemon** [comment]
@TemoTech25 Please leave the repository link so we can see the code.

**Reply by TemoTech25** [reply]
/home/adduser/solidity-course/foundry-f23/foundry-fund-me-f23/.git/

**Reply by TemoTech25** [reply]
![Screenshot (5)](https://github.com/Cyfrin/foundry-full-course-f23/assets/135317073/ed396dfb-9131-44fb-a22f-722cd575947e)
![Screenshot (6)](https://github.com/Cyfrin/foundry-full-course-f23/assets/135317073/eef0dc15-4c00-4d76-8b06-66134690fd55)
![Screenshot (7)](https://github.com/Cyfrin/foundry-full-course-f23/assets/135317073/b4154eff-e8d3-46e0-b855-81e16986bf2b)

**Reply by cromewar** [reply]
@TemoTech25 
`/home/adduser/solidity-course/foundry-f23/foundry-fund-me-f23/.git/`
This is your local repo, but please upload this to Github and send the link so we can check it.

**Reply by TemoTech25** [reply]
https://github.com/TemoTech25/my-learning-foundry-FundMe-project-

**Reply by TemoTech25** [reply]
Is Anyone able assist me

**Reply by ghost** [reply]
@TemoTech25 check below for wannabeOreki answer.

**Comment by jobmwitah** [comment]
Can we see your tests?

**Reply by TemoTech25** [reply]
which tests?

**Reply by TemoTech25** [reply]
https://github.com/TemoTech25/my-learning-foundry-FundMe-project-

**Comment by wannabeOreki** [comment, marked as answer]
For testing, are you forking from the Sepolia chain? Because from the above code snippet, I don't see any code written for the deployment of `MockV3Aggregator` from the local anvil chain.
Use the below command instead (Note: here I am assuming you have successfully created a .env file and pasted your SEPOLIA_RPC_URL in it, if not then checkout this [discussion](https://github.com/Cyfrin/foundry-full-course-f23/discussions/107) and [Private Key Rant II](https://youtu.be/umepbfKp5rI?t=27701) )

``` shell
forge test --fork-url $SEPOLIA_RPC_URL
```
Since the local anvil chain doesn't contain `AggregatorV3Interface`, we have to create a mock of `AggregatorV3Interface`.You can deploy your mock in the test directory like this.
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/132799458/401f12e5-8879-477e-8795-022435b49260)

Checkout [Mocking-contracts](https://youtu.be/sas02qSFZ74?t=3645), here @PatrickAlphaC has explained why and when we need to deploy mocks

Once you deploy a `MockV3Aggregator`, modify your `getAnvilEthConfig` function, in such a way that it deploys a new mock every time it gets called and stores the address of this mock into `anvilNetworkConfig`. Below is the code written for that.
```solidity
function getAnvilEthConfig()
        public
        returns (NetworkConfig memory anvilNetworkConfig)
    {
        if (anvilNetworkConfig.priceFeed != address(0)) {
            return anvilNetworkConfig;
        }
        vm.startBroadcast();
        MockV3Aggregator mockPriceFeed = new MockV3Aggregator(
            DECIMALS,
            INITIAL_PRICE
        );
        vm.stopBroadcast();
        emit HelperConfig__CreatedMockPriceFeed(address(mockPriceFeed));
        anvilNetworkConfig = NetworkConfig({priceFeed: address(mockPriceFeed)});
    }
```
I hope this solves the issue!
BTW this was my first ever involvement in any discussions, so feel free to ask me if I was not able to explain things properly

**Reply by Equious** [reply]
You did great for a first contribution! Thank you for your answer!

**Comment by DevSwayam** [comment]
@TemoTech25 You got the answer?

**Comment by PatrickAlphaC** [comment]
Hello! Thanks for making this discussion. 

Could you please remember to follow this section to format your questions? https://youtu.be/umepbfKp5rI?t=22464
